### PR TITLE
PC, paulse elements, A/S and shared memory objects wrappers for the ZIS AUX projects.

### PR DESCRIPTION
**Pull request by ifakhrutdinov**
The following wrappers have been added:
* Address space management
* Pause elements
* Program calls
* 64-bit shared memory objects

### CODE REVIEW

File: c/shrmem64.c [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+#error ILP32 is not supported
+#endif
+
+typedef uint64_t MemObj;
+
+#define IARV64_V4PLIST_SIZE 160
+
+#define IS_IARV64_OK($rc) (($rc) < 8)
+
+#define MAKE_RSN($functionRC, $iarv64RC, $iarv64RSN) \
```
Comment (daveyc) [inline c/shrmem64.c]: Prefer functions to macros

File: c/shrmem64.c [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+#ifndef _LP64
+#error ILP32 is not supported
+#endif
+
+typedef uint64_t MemObj;
+
+#define IARV64_V4PLIST_SIZE 160
+
+#define IS_IARV64_OK($rc) (($rc) < 8)
```
Comment (daveyc) [inline c/shrmem64.c]: Prefer functions to macros

File: c/shrmem64.c [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+#include <shrmem64.h>
+#include "zos.h"
+
+#ifndef _LP64
+#error ILP32 is not supported
+#endif
+
+typedef uint64_t MemObj;
+
+#define IARV64_V4PLIST_SIZE 160
```
Comment (daveyc) [inline c/shrmem64.c]: prefer "static const" to #define

File: c/shrmem64.c [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+  return result.token;
+}
+
+int shrmem64Alloc(MemObjToken userToken, size_t size, void **result, int *rsn) {
+
+  uint32_t iarv64RC = 0, iarv64RSN = 0;
+
+  uint64_t segmentCount = 0;
+  if ((size & 0xFFFFF) == 0){
+    segmentCount = size >> 20;
```
Comment (daveyc) [inline c/shrmem64.c]: Please simplify or explain the mask with a comment or constant.

File: c/shrmem64.c [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+  } else{
+    segmentCount = (size >> 20) + 1;
```
Comment (daveyc) [inline c/shrmem64.c]: Ditto

File: c/shrmem64.c (hunk 3 above)
Comment (ifakhrutdinov) [reply c/shrmem64.c]: I'm not sure I want to use a static const. That value is used to define the size of an array. A static const would make the compiler use VLA which are optional in C11.

**Pull request merged**